fpsd/features.py

```diff
@@ -0,0 +1,768 @@
+            panic("FingerprintSecureDrop Postgres support relies on use of a "
+                  "PGPASSFILE. Make sure this file and the env var pointing "
+                  "to it exist and set 0600 permissions & user ownership."
+                  "\n{}.".format(exc))
+
+    def drop_table(self, table_name):
+        """Try to remove a table even if views depend on it
+
+        Args:
+            table_name [string]: Name of table to be dropped
+        """
+
+        query = "DROP TABLE IF EXISTS {} CASCADE;".format(table_name)
+        self.engine.execute(query)
+
+    def generate_table_undefended_frontpage_links(self):
+        """This method creates a table of exampleids that were present 
+        at the time of feature generation that the unified view can join
+        on. The table contains one integer column, exampleid.
+        """
+
+        self.drop_table("features.undefended_frontpage_examples")
+
+        query = ("CREATE TABLE features.undefended_frontpage_examples AS ( "
+                 "SELECT foo.exampleid FROM ( SELECT exampleid, "
+                 "count(*) FROM raw.frontpage_traces GROUP BY exampleid) foo);")
+        self.engine.execute(query)
+
+    def _create_temp_packet_positions(self, outgoing_only=True):
+        """This method takes all rows in raw.frontpage_traces and
+        creates a temporary table packet_positions with the following
+        format:
+
+        exampleid  | position |  ingoing
+        (integer)  | (bigint) |  (boolean)
+        ------------------------------------
+        9          | 4        | f
+        9          | 5        | f
+        9          | 15       | f
+
+        where position describes where the cell is in the trace. For
+        example, for the first outgoing cell in the trace with
+        exampleid=9 was the 4th packet in the trace. This table is used
+        by the position-based feature generation functions.
+
+        Args:
+            outgoing_only [boolean]: describes whether or not we should
+                only use outgoing cells to create this table
+        """
+
+        self.drop_table("packet_positions")
+
+        if outgoing_only:
+            where_only_outgoing = "WHERE ingoing <= false"
+        else:
+            where_only_outgoing = ""
+
+        query = ("CREATE TEMP TABLE packet_positions AS             "
+                 "(SELECT exampleid, position, ingoing              "
+                 "  FROM (                                          "
+                 "    SELECT                                        "
+                 "      ROW_NUMBER() OVER                           "
+                 "      (PARTITION BY exampleid ORDER BY t_trace)   "
+                 "      AS position,                                "
+                 "      t.*                                         "
+                 "    FROM raw.frontpage_traces t)                  "
+                 "x {} );").format(where_only_outgoing)
```

Comment:
The subquery is necessary here - this won’t work where we are selecting only the rows where `ingoing = false`